src/clr-angular/data/tree-view/models/recursive-tree-node-model.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { TreeNodeModel } from './tree-node-model';
+import { AsyncArray, isPromise } from './async-array';
+import { isObservable, Subscription } from 'rxjs';
+
+/*
+ * A recursive model is built received from the app and traversed to create the corresponding components.
+ * Recursive = Model dictates the tree node components
+ */
+export class RecursiveTreeNodeModel<T> extends TreeNodeModel<T> {
+  constructor(
+    model: T,
+    parent: RecursiveTreeNodeModel<T> | null,
+    private getChildren: (node: T) => AsyncArray<T> | undefined
+  ) {
+    super();
+    this.model = model;
+    this.parent = parent;
+  }
+
+  parent: RecursiveTreeNodeModel<T> | null;
+
+  private childrenFetched = false;
+
+  fetchChildren() {
+    if (this.childrenFetched) {
+      return;
+    }
+
+    const asyncChildren = this.getChildren(this.model);
+    if (isPromise(asyncChildren)) {
+      this.loading = true;
+      asyncChildren.then(raw => {
+        this._children = this.wrapChildren(raw);
+        this.loading = false;
+      });
+    } else if (isObservable(asyncChildren)) {
+      this.loading = true;
+      this.subscription = asyncChildren.subscribe(raw => {
+        this._children = this.wrapChildren(raw);
+        this.loading = false;
+      });
+    } else if (asyncChildren) {
+      // Synchronous case
+      this._children = this.wrapChildren(asyncChildren);
+    } else {
+      this._children = [];
+    }
+    this.childrenFetched = true;
```

Comment:
What if they want to re-fetch the children for a given node (perhaps the backend refreshed)?

Reply:
That's a greater question. 😛 

Because of time constraints, I considered they could refresh the tree itself if children have changed. Not ideal obviously, but not a blocker as far as I could tell. I believe the next feature we could add is that if the children received are an Observable, we keep up to date with it. That would be a way for them to push dynamic changes.

Are you okay with us moving forward without it for now? If yes I'll create an issue that we can address in 1.x.

Reply:
I'm ok with this being a future enhancement, doesn't sound like a breaking change.